### ISSUE TITLE
Feature Request: highlight selected word occurances

### ISSUE THREAD

**Issue by GitHubIsToxic** [linuxmint/xed#217]
Since gedit/xed is often used to edit code or configuration files, it would benefit from a feature to highlight selected word occurrences. I personally find this feature indispensable in any plain text editor nowadays. Every text editor I'm aware of has this feature.

Perhaps port Smart Highlighting plugin (https://github.com/nemec/gedit-smart-highlight-plugin) and include it with Xed, as a stopgap solution? I would also not mind seeing the plugin enabled by default.

Although perhaps it would be better to implement this feature natively, and make it an option in preferences.

**Comment by GitHubIsToxic** [comment]
I tried implementing this myself, but it turned out very difficult, because Xed code is almost entirely devoid of comments (see #252).

Can someone familiar with the Xed source code point me to some code/locations that can help me?
Would it be easier to implement as plugin, or better to integrate with core code?
Or perhaps implementation is so trivial that might less time consuming to implement this feature rather than comment code and/or to explain it to me? Can I assume the document gtk widget has a word occurrence highlight feature? I looked at the existing text search code, but it's hard to make sense of things without comments.

Suggestions are welcome.

**Comment by GitHubIsToxic** [comment]
By the way, for users faced with the same need of this feature, I would recommend Kate (or Kedit, a simpler version of Kate), which has this standard feature.

**Comment by tpamula** [comment]
From what I've noticed such a feature already exists in xed but has to be invoked manually. If you higlight a word and enter the search via ctrl+f or by clicking the search icon it highlights all the other occurrences by default. 

It's not the same but pretty similar. Even if it doesn't suit your needs, you could have a look at the implementation of this feature - it might bring some inspiration.

**Comment by GitHubIsToxic** [comment]
Yes, the gedit 2 plugin I referenced does exactly that, and I just happened to have implemented the exact same feature in another project this week. It should be fairly trivial (unless weird APIs or bugs get in the way...not uncommon).

But again, the Xed code is devoid of comments, so it's hard to even find the code related to regular search function, let alone plug in some code that piggy-backs on that.
I believe the Mint maintainers ran some C automated reformatting scripts to clean up and re-structure the old gedit code when they forked Xed, but that's also wiped out a lot of (potentially) useful hints. That said (and having worked with decades old of incorrectly and outdated commented X11 and Gnome code), it may equally have been a good decision to lose the comments.

I'm just wishing for someone more intimate with the gedit/xed code to sprinkle a few comments here and there to get things started (and become more inviting to tinkering). For example a summary at the top of each source file which functionality can be found below would be a good start.

**Comment by collinss** [comment]
This is now implemented. Closing...

**Comment by GitHubIsToxic** [comment]
Great, many thanks!

@collinss: could you point me to the commit that implements this? I'm still interested in the code (as mentioned in #252), but I can't find it. According to comments on StackOverflow, Github strangely doesn't seem to have commit log feature that can show a SVN-style overview of commits.

**Comment by collinss** [comment]
Hmm... I just reread your initial request, and apparently I misunderstood what you were asking for. Now I feel dumb. This feature actually hasn't been implemented. Reopening...

**Comment by GitHubIsToxic** [comment]
Oh bummer. But thanks anyway.

**Comment by Primer81** [comment]
For the record, I would hazard a guess that the relevant function for highlighting text occurrences would be this function in `xed-document.c`:

```
void
xed_document_set_search_context (XedDocument            *doc,
                                 GtkSourceSearchContext *search_context)
{
    XedDocumentPrivate *priv;

    g_return_if_fail (XED_IS_DOCUMENT (doc));

    priv = xed_document_get_instance_private (doc);

    g_clear_object (&priv->search_context);
    priv->search_context = search_context;

    if (search_context != NULL)
    {
        gboolean highlight = g_settings_get_boolean (priv->editor_settings, XED_SETTINGS_SEARCH_HIGHLIGHTING);

        gtk_source_search_context_set_highlight (search_context, highlight);

        g_object_ref (search_context);
    }
}
```

Specifically the call to `gtk_source_search_context_set_highlight (search_context, highlight);`. I only say this because the other function named `_xed_cmd_search_clear_highlight (XedWindow *window)` from `xed-commands-search.c` calls this function to remove the highlight.

Not too familiar with the code base though so I could be wrong.

**Comment by gabriellaraujocoding** [comment]
Hi. I published a drop-in Xed plugin that does what this request describes: **automatically highlight all occurrences of the current selection** (no Ctrl+F needed), without recompiling Xed.

**xed-quick-highlight** plugin:
- Highlights all matches of the selected text
- Works only for non-empty, single-line selections (avoids expensive multi-line patterns)
- Updates automatically while you move the caret or edit

Plugin: https://github.com/gabriellaraujocoding/xed-extra-plugins/tree/main/xed-quick-highlight
Repository: https://github.com/gabriellaraujocoding/xed-extra-plugins

Screenshot:
![xed-quick-highlight](https://raw.githubusercontent.com/gabriellaraujocoding/xed-extra-plugins/main/screenshots/xed-quick-highlight.png)

Install (no rebuild):
```bash
mkdir -p ~/.local/share/xed/plugins/
cp -r xed-quick-highlight ~/.local/share/xed/plugins/
```
Enable: **Edit → Preferences → Plugins → Xed Quick Highlight**
